### PR TITLE
Add tests/matchers for automatic accessibility testing

### PR DESCRIPTION
**Pull request by jonahwilliams**
Based on guidelines from WCAG and the Android Accessibility Scanner, these new `Guideline`s verify that the results of `test.pumpWidget()` meet some accessibility guidelines.  This change includes two of these guidelines by default:

- (android/iOS)TapTargetSizeGuideline, which verifies nodes with tap/long press events are at least 48x48/44x44

- textContrastGuideline, verifies that any text which contributes semantics meets the minimum contrast guidelines established in WCAG 2.0 (https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html)

EDIT:

The algorithm for contrast is fairly trivial - @gspencergoog and I discussed some possible improvements offline, but they will require some research and experimentation.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
cc @goderbauer  as well

File: packages/flutter_test/lib/src/accessibility.dart [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+  FutureOr<Evaluation> evaluate(WidgetTester tester);
+
+  /// A description of the policy restrictions and criteria.
+  String get description;
+}
+
+class _MinimumTapTargetGuideline extends AccessibilityGuideline {
+  const _MinimumTapTargetGuideline(this.size);
+
+  final Size size;
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/accessibility.dart]: nit: maybe document that this is the expected minimum size?

File: packages/flutter_test/lib/src/accessibility.dart [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+      }
+      final List<int> subset = _subsetToRect(byteData, paintBounds, image.width, image.height);
+      final _ContrastReport report = new _ContrastReport(subset);
+      final double contrastRatio = report.contrastRatio();
+      final double targetContrastRatio = (isBold && fontSize > kBoldTextMinimumSize) ?
+        kMinimumRatioLargeText : kMinimumRatioNormalText;
+      if (contrastRatio >= targetContrastRatio)
+        return result + const Evaluation.pass();
+      return result + new Evaluation.fail(
+        '$node:\nExpected contrast ratio of at least '
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/accessibility.dart]: Would it make sense to include a link to why these are the recommended contrasts in the reason? e.g. the links contained in the doc comments above?

(same for the failure reason of the other tester)

File: packages/flutter_test/lib/src/matchers.dart [hunk 3]
```diff
@@ -457,6 +459,36 @@ Matcher matchesSemanticsData({
...
+/// `guideline`.
+///
+/// This matcher requires the result to be awaited and for semantics to be
+/// enabled first.
+///
+/// ## Sample code
+///
+/// ```dart
+/// final SemanticsHandle handle = tester.ensureSemantics();
+/// await meetsGuideline(tester, meetsGuideline(textContrastGuideline));
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/matchers.dart]: include disposing the handle in this example?

File: packages/flutter_test/lib/src/matchers.dart [hunk 4]
```diff
@@ -457,6 +459,36 @@ Matcher matchesSemanticsData({
...
+/// ```
+///
+/// Supported accessibility guidelines:
+///
+///   * [androidTapTargetGuideline]
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/matchers.dart]: nit: include a one line description of what these do coma separated from the link.

File: packages/flutter_test/lib/src/matchers.dart [hunk 5]
```diff
@@ -457,6 +459,36 @@ Matcher matchesSemanticsData({
...
+///   * [iOSTapTargetGuideline]
+///   * [textContrastGuideline]
+AsyncMatcher meetsGuideline(AccessibilityGuideline guideline) {
+  return new _MatchesAccessibilityGuideline(guideline);
+}
+
+/// The inverse matcher of [meetsGuideline].
+///
+/// This is needed because the [isNot] matcher does not compose with an
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/matchers.dart]: sad. Maybe there should be an upstream isNotAsync matcher?

File: packages/flutter_test/test/accessibility_test.dart [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+  group('text contrast guideline', () {
+    testWidgets('black text on white background - Text Widget - direct style', (WidgetTester tester) async {
+      final SemanticsHandle handle = tester.ensureSemantics();
+      await tester.pumpWidget(_boilerplate(
+        const Text(
+          'this is a test',
+          style: TextStyle(fontSize: 14.0, color: Colors.black),
+        ),
+      ));
+      await expectLater(tester, meetsGuideline(textContrastGuideline));
```
Comment (goderbauer) [inline packages/flutter_test/test/accessibility_test.dart]: Can you add a test that ensures that a failing matcher actually produces the expected errors?

**Review by goderbauer — APPROVED** [review]
LGTM

This is awesome!

File: packages/flutter_test/lib/src/matchers.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_test/lib/src/matchers.dart]: There are, but they don't really mesh with how we handle async in widget tests, hence the AsyncMatcher work around.  I'm not really sure what the solution is here, maybe just make the guideline classes public so they can be unit tested?

File: packages/flutter_test/test/accessibility_test.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter_test/test/accessibility_test.dart]: Same issue as above

File: packages/flutter_test/lib/src/matchers.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_test/lib/src/matchers.dart]: Done

File: packages/flutter_test/lib/src/matchers.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_test/lib/src/matchers.dart]: Done

File: packages/flutter_test/lib/src/accessibility.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_test/lib/src/accessibility.dart]: Done

File: packages/flutter_test/lib/src/accessibility.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_test/lib/src/accessibility.dart]: Done

File: packages/flutter_test/lib/src/matchers.dart (hunk 5 above)
Comment (Hixie) [reply packages/flutter_test/lib/src/matchers.dart]: is there a bug filed on the `test` package about this limitation?

File: packages/flutter_test/test/accessibility_test.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter_test/test/accessibility_test.dart]: Made the guideline classes (but not constructors) public so they can be unit tested, add failure message test cases.

File: packages/flutter_test/lib/src/matchers.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_test/lib/src/matchers.dart]: Filled https://github.com/dart-lang/test/issues/923

**Comment by Hixie** [comment]
This PR is amazing.

**Pull request merged**